optuna/samplers/_tpe/multivariate_parzen_estimator.py

```diff
@@ -0,0 +1,422 @@
+                transformed[param_name] = np.clip(samples, low, high)
+            elif isinstance(distribution, distributions.IntUniformDistribution):
+                q = self._q[param_name]
+                samples = np.round(samples / q) * q
+                samples = np.clip(samples, self._low[param_name], self._high[param_name])
+                transformed[param_name] = samples.astype(int)
+            elif isinstance(distribution, distributions.IntLogUniformDistribution):
+                transformed[param_name] = np.round(np.exp(samples))
+            elif isinstance(distribution, distributions.CategoricalDistribution):
+                transformed[param_name] = samples
+            else:
+                distribution_list = [
+                    distributions.UniformDistribution.__name__,
+                    distributions.LogUniformDistribution.__name__,
+                    distributions.DiscreteUniformDistribution.__name__,
+                    distributions.IntUniformDistribution.__name__,
+                    distributions.IntLogUniformDistribution.__name__,
+                    distributions.CategoricalDistribution.__name__,
+                ]
+                raise NotImplementedError(
+                    "The distribution {} is not implemented. "
+                    "The parameter distribution should be one of the {}".format(
+                        distribution, distribution_list
+                    )
+                )
+        return transformed
+
+    def _precompute_sigmas0(
+        self, multivariate_samples: Dict[str, np.ndarray]
+    ) -> Optional[Union[np.ndarray, float]]:
+
+        # Categorical parameters are not considered.
+        rescaled_samples_list = []
+        for param_name, param_dist in self._search_space.items():
+            if isinstance(param_dist, _NUMERICAL_DISTRIBUTION_CLASSES):
+                high = self._high[param_name]
+                low = self._low[param_name]
+                assert high is not None
+                assert low is not None
+                samples = multivariate_samples[param_name]
+                samples = (samples - low) / (high - low)
+                rescaled_samples_list.append(samples)
+            else:  # Categorical parameters are ignored.
+                continue
+
+        # When the number of parameters is zero, we cannot determine sigma0.
+        # If there are only categorical parameters, this case happens.
+        if len(rescaled_samples_list) == 0:
+            return None
+        # When the number of samples is zero, we return 1.0.
+        elif len(rescaled_samples_list[0]) == 0:
+            return 1.0
+
+        rescaled_samples = np.array(rescaled_samples_list).T
+
+        # compute distance matrix of samples
+        distances = np.linalg.norm(
+            rescaled_samples[:, None, :] - rescaled_samples[None, :, :], axis=2
+        )
+        distances[np.diag_indices_from(distances)] += np.inf
+
+        return np.min(distances, axis=1)
```

Comment:
### Suggestion of new bandwidth selection rule (Scott's rule)

The current implementation of bandwidth selection uses the distance from the nearest neighbor sample as the bandwidth. However, this rule requires O(d*n^2) complexity for n samples with d parameters.

Thus, I suggest the use of another rule called Scott's rule, which is of complexity O(1). This rule sets the bandwidth proportionally to n^{-1/(d+4)}, which is theoretically optimal order of the bandwidth. 

In my benchmark (shown in the attached file), there is no significant difference between current implementation and Scott's rule. Also in the BOHB paper, Scott's rule is used for bandwidth selection in combination with multivariate TPE.

Please let me know your opinion on this.
```suggestion
    def _precompute_sigmas0(self, multivariate_samples: Dict[str, np.ndarray]) -> float:
        sample_size = next(iter(multivariate_samples.values())).size
        param_size = len(multivariate_samples)
        return sample_size ** (-1.0 / (param_size + 4))
```
Also, we can get rid of the following at line 278.
```
assert sigmas0 is not None
```


![hpo-bench-naval-40f361d00195ba642c8866791ccec1b37a5cb8f0e3b1cf3092f76d731d4f9033](https://user-images.githubusercontent.com/30857855/92066478-f0fe3100-eddc-11ea-9e40-4baf81ab8849.png)
![hpo-bench-parkinson-e042118c285061b2da9d3a89fc6c1dbf4b7f2f8384e8db7cabae2f31a5c2842e](https://user-images.githubusercontent.com/30857855/92066480-f22f5e00-eddc-11ea-954d-746e71cc048e.png)
![hpo-bench-protein-aad4c2d81b4340ee2dc4ee252cea7eea3a49a3478fdbdf6065c265f99cd522f0](https://user-images.githubusercontent.com/30857855/92066483-f2c7f480-eddc-11ea-89d8-2cae0121047f.png)
![hpo-bench-slice-8ab21f97e472c9f9120602a69877fad413f2bacdae10ec7d1b637303fb04cbcc](https://user-images.githubusercontent.com/30857855/92066485-f3608b00-eddc-11ea-9da6-8ed3d4293246.png)

Reply:
Looks great! Using Scott's rule seems reasonable.

Reply:
By the way, in the BOHB paper, they multiply the bandwidths with a constant factor to promote exploration (Section 4.1. in https://arxiv.org/pdf/1807.01774.pdf). Is this something that has been considered?

Reply:
OK, I will adopt Scott's rule, with some comments for future maintenance.

Reply:
I am sorry to announce that I made a mistake in the above benchmark. 
I will not adopt this at the moment and will close this thread for now. If it performs well in the fixed benchmark, I might reopen this thread.

Reply:
> By the way, in the BOHB paper, they multiply the bandwidths with a constant factor to promote exploration (Section 4.1. in https://arxiv.org/pdf/1807.01774.pdf). Is this something that has been considered?

No, but I believe this is worth exploring.

Reply:
How about leaving a TODO?

Reply:
I took the benchmark again, and fortunately, Scott's rule outperformed the current strategy.
So I modified `_precompute_sigmas0` and added comments on TODOs.